File: medium/207. Course Schedule/207. Course Schedule.ts
// Possible solve
// 1. Create graph in "adjacency list" way
// 2. For each course in graph fill an array with pre visited courses
// 3. State will help control the visit status for each course: 0 - NOT visited, 1 - visiting, 2 - completed
// 4. Create DFS function to find a cycle: control visited status in state
// 5. Run DFS function for each course and try to find cycle

function canFinish(numCourses: number, prerequisites: number[][]): boolean {
    // Build a graph
    const graph: { [key in number]: number[] } = Object.fromEntries(
        Array.from({ length: numCourses }, (_, i) => [i, []])
    )

    for (const [course, pre] of prerequisites) {
        graph[course].push(pre)
    }

    // State for each course: 0 - NOT visited, 1 - visiting, 2 - completed
    const state = new Array(numCourses).fill(0)

    // DFS function to find cycle in graph, if it exists return true, otherwise - false
    const hasCycle = (course: number): boolean => {
        if (state[course] == 1)
            return true
        else if (state[course] == 2)
            return false

        state[course] = 1

        for (const pre of graph[course]) {
            if (hasCycle(pre))
                return true
        }

        state[course] = 2
        return false

    }

    // Check each course to find a cycle
    for (let course = 0; course < numCourses; course++) {
        const isCycle = hasCycle(course)
        if (isCycle)
            return false
    }

    return true
};


console.log(canFinish(2, [[1, 0]]))
console.log(canFinish(2, [[1, 0], [0, 1]]))

export { }

File: medium/207. Course Schedule/207. Course Schedule.py
# from collections import defaultdict
from typing import List


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = {i: [] for i in range(numCourses)}
        for course, pre in prerequisites:
            graph[course].append(pre)

        state = [0] * numCourses

        def has_cycle(course: int) -> bool:
            if state[course] == 1:
                return True
            if state[course] == 2:
                return False

            state[course] = 1
            for pre in graph[course]:
                if has_cycle(pre):
                    return True

            state[course] = 2
            return False

        for course in range(numCourses):
            if has_cycle(course):
                return False

        return True


solution = Solution()
# print(solution.canFinish(2, [[1, 0]]))
print(solution.canFinish(2, [[1, 0], [0, 1]]))
print(solution.canFinish(3, [[1, 0], [2, 1]]))
